`TagCompound` arrays

ExterminatorX99 opened the pull request:
### What is the new feature?
Support for serializing arrays into `TagCompound`

<ins>NBT spec compatible</ins>

### Why should this be part of tModLoader?
Currently if modders want to save arrays of data they have to first convert it to a list and only then save it.
Same with loading. They have to load the data as a list, then convert it to an array.

### Are there alternative designs?
Current. Only support for `byte` and `int` arrays

### Sample usage for the new feature / What I used for testing
```cs
public override void SaveWorldData(TagCompound tag) {
	tag["intArr1"] = new int[] { 0, 1, 2, 3 };
	tag["intArr2"] = new int[,] { { 0, 1 }, { 2, 3 } };
	tag["intArr3"] = new int[,,] { { { 0, 1 }, { 2, 3 } }, { { 4, 5 }, { 6, 7 } } };
	tag["intArr4"] = new int[,,,] { { { { 0, 1 }, { 2, 3 } }, { { 4, 5 }, { 6, 7 } } }, { { { 8, 9 }, { 10, 11 } }, { { 12, 13 }, { 14, 15 } } } };

	tag["intArr2Arr2"] = new int[,][,] { { new[,]{ { 0, 1 }, { 2, 3 } }, new[,]{ { 4, 5 }, { 6, 7 } } }, { new[,]{ { 8, 9 }, { 10, 11 } }, new[,]{ { 12, 13 }, { 14, 15 } } } };

	tag["intListArr"] = new List<int>[] { new List<int> { 0, 1, 2, 3 }, new List<int> { 4, 5, 6, 7 } };
	tag["intArrList"] = new List<int[]> { new int[] { 0, 1, 2, 3 }, new int[] { 4, 5, 6, 7 } };

	tag["vectorArr"] = new Vector2[] { new Vector2(0, 1), new Vector2(2, 3) };
	tag["vectorList"] = new List<Vector2> { new Vector2(0, 1), new Vector2(2, 3) };
	tag["vectorListArr"] = new List<Vector2[]> { new Vector2[] { new Vector2(0, 1), new Vector2(2, 3) }, new Vector2[] { new Vector2(4, 5), new Vector2(6, 7) } };
	tag["vectorArrList"] = new List<Vector2>[] { new List<Vector2> { new Vector2(0, 1), new Vector2(2, 3) }, new List<Vector2> { new Vector2(4, 5), new Vector2(6, 7) } };

	tag["vectorArr2Arr2"] = new Vector2[,][,] { { new Vector2[,] { { new Vector2(0, 1), new Vector2(2, 3) }, { new Vector2(4, 5), new Vector2(6, 7) } } }, { new Vector2[,] { { new Vector2(8, 9), new Vector2(10, 11) }, { new Vector2(12, 13), new Vector2(14, 15) } } } };
	tag["vectorArr2Arr2List"] = new List<Vector2[,][,]> { new Vector2[,][,] { { new Vector2[,] { { new Vector2(0, 1), new Vector2(2, 3) }, { new Vector2(4, 5), new Vector2(6, 7) } } }, { new Vector2[,] { { new Vector2(8, 9), new Vector2(10, 11) }, { new Vector2(12, 13), new Vector2(14, 15) } } } }, { new Vector2[,][,] { { new Vector2[,] { { new Vector2(0, 1), new Vector2(2, 3) }, { new Vector2(4, 5), new Vector2(6, 7) } } }, { new Vector2[,] { { new Vector2(8, 9), new Vector2(10, 11) }, { new Vector2(12, 13), new Vector2(14, 15) } } } } } };
	tag["vectorListArr2Arr2"] = new List<Vector2>[,][,] { { new List<Vector2>[,] { { new List<Vector2>() { new Vector2(0, 1), new Vector2(2, 3) } }, { new List<Vector2>() { new Vector2(4, 5), new Vector2(6, 7) } } } }, { new List<Vector2>[,] { { new List<Vector2>() { new Vector2(8, 9), new Vector2(10, 11) } }, { new List<Vector2>() { new Vector2(12, 13), new Vector2(14, 15) } } } } };

	tag["array"] = new Vector2[] { new Vector2(0, 1), new Vector2(2, 3) };
	tag["list"] = new List<Vector2> { new Vector2(0, 1), new Vector2(2, 3) };
}

public override void LoadWorldData(TagCompound tag) {
	var list = tag.Get<List<int>>("doesNotExist");
	var array = tag.Get<int[,]>("doesNotExist");

	var intArr1 = tag.Get<int[]>("intArr1");
	var intArr2 = tag.Get<int[,]>("intArr2");
	var intArr3 = tag.Get<int[,,]>("intArr3");
	var intArr4 = tag.Get<int[,,,]>("intArr4");

	var intArr2Arr2 = tag.Get<int[,][,]>("intArr2Arr2");

	var intListArr = tag.Get<List<int>[]>("intListArr");
	var intArrList = tag.Get<List<int[]>>("intArrList");

	var vectorArr = tag.Get<Vector2[]>("vectorArr");
	var vectorList = tag.Get<List<Vector2>>("vectorList");
	var vectorListArr = tag.Get<List<Vector2[]>>("vectorListArr");
	var vectorArrList = tag.Get<List<Vector2>[]>("vectorArrList");

	var vectorArr2Arr2 = tag.Get<Vector2[,][,]>("vectorArr2Arr2");
	var vectorArr2Arr2List = tag.Get<List<Vector2[,][,]>>("vectorArr2Arr2List");
	var vectorListArr2Arr2 = tag.Get<List<Vector2>[,][,]>("vectorListArr2Arr2");

	var arrayToList = tag.Get<List<Vector2>>("array");
	var listToArray = tag.Get<Vector2[]>("list");
}
```

### ExampleMod updates
`ExampleInstancedItem` remove conversion between array and list

ExterminatorX99 committed:
Array serialization

ExterminatorX99 committed:
Array deserialization

ExterminatorX99 commented on patches/tModLoader/Terraria/ModLoader/IO/TagIO.cs:
```diff
@@ -118,7 +122,7 @@ private class ClassPayloadHandler<T> : PayloadHandler<T> where T : class
 						return GetHandler(GetPayloadId(v.GetType().GetGenericArguments()[0])).CloneList(v);
 					}
 					catch (IOException) {
-						throw new ArgumentException("Invalid NBT list type: "+v.GetType());
```
Is there some good reason this was an `ArgumentException` instead of an `IOException`?

Chicken-Bones replied:
This one is `ArgumentException` because the IO is fine, you've just requested the wrong type for the data, so the type _argument_ is invalid. The same logic could apply to the one above, for writing, but I guess it doesn't matter either way. I'm fine with the change to `IOException`

ExterminatorX99 committed:
EM usage

Chicken-Bones commented:
I'd like you to do this without the binary breaking change. Supporting tools like `NBTExplorer` is fairly important to us. In fact, 12 is actually used by long arrays now! (and we should support it, why not).

I actually don't know why NBT even bothered with array types, since they don't appear to be any more efficient than the equivalent primitive list.

absoluteAquarian commented:
I agree with Chicken-Bones.  NBTExplorer is such a useful tool for checking the data in `.twld` files to the point that stepping beyond the NBT specifications would cause it to become potentially useless.

Using `.ToList()` and `.ToArray()` already covers "using arrays" in NBT data.

ExterminatorX99 committed:
NBT spec compatibility

ExterminatorX99 committed:
Organize + cleanup

ExterminatorX99 committed:
public

ExterminatorX99 committed:
Rename method

ExterminatorX99 commented:
Has bug. Don't merge yet

ExterminatorX99 committed:
Fix typo bug

ExterminatorX99 committed:
Null guards

ExterminatorX99 committed:
Remove faulty early return

ExterminatorX99 committed:
Change rank length tag id

ExterminatorX99 committed:
Allow null converter with `FromTagCompound`

ExterminatorX99 committed:
Rename some things

ExterminatorX99 committed:
Fix issues with returning default arrays

ExterminatorX99 committed:
Merge remote-tracking branch 'upstream/1.4' into 1.4_tag_array

ExterminatorX99 committed:
Make lists and 1d arrays interchangeable

ExterminatorX99 committed:
Remove unused method

ExterminatorX99 committed:
Convert to `TagSerializer`

ExterminatorX99 committed:
Merge remote-tracking branch 'upstream/1.4' into 1.4_tag_array

ExterminatorX99 committed:
Use `Activator` instead of being manual

ExterminatorX99 committed:
Make `TagSerializer` nullable as a side dish

ExterminatorX99 committed:
Some safety

ExterminatorX99 committed:
Fix dumb mistake

ExterminatorX99 committed:
Cleanup

Move empty array creation into serializer
Revert some list of list serialization changes

ExterminatorX99 committed:
Misc changes

ExterminatorX99 committed:
Merge remote-tracking branch 'upstream/1.4' into 1.4_tag_array

ExterminatorX99 committed:
`TagIO` nullable enable

ExterminatorX99 committed:
Add array/list interchangeability tests

ExterminatorX99 committed:
Add multi dimensional array tests

ExterminatorX99 committed:
Fix for multi dim test failing

ExterminatorX99 committed:
Add default value tests for multi dim arrays

ExterminatorX99 committed:
Add tests for arrays of lists and lists of arrays

ExterminatorX99 committed:
Fix tests for arrays of lists and lists of arrays

ExterminatorX99 committed:
Duck typing tests for multi dim arrays

ExterminatorX99 committed:
Merge remote-tracking branch 'upstream/1.4' into 1.4_tag_array

ExterminatorX99 committed:
Formatting

ExterminatorX99 committed:
More duck typing tests

ExterminatorX99 committed:
We don't need to store the serialized type

ExterminatorX99 committed:
Fix issues returning empty multi dim arrays

ExterminatorX99 committed:
Rename some stuff for clarity

ExterminatorX99 committed:
Slightly reorganize

ExterminatorX99 committed:
Formatting

ExterminatorX99 committed:
Clearer error messages

ExterminatorX99 committed:
`TagCompound.TryGet`

ExterminatorX99 committed:
Merge remote-tracking branch 'upstream/1.4' into 1.4_tag_array

ExterminatorX99 committed:
Merge remote-tracking branch 'upstream/1.4' into 1.4_tag_array

ExterminatorX99 committed:
Simplify the way array ranks are stored

ExterminatorX99 committed:
Extract constant strings

ExterminatorX99 committed:
Comment and slight optimization
